src/main/java/org/apache/sysds/runtime/instructions/fed/MultiReturnParameterizedBuiltinFEDInstruction.java

```diff
@@ -0,0 +1,215 @@
+import org.apache.sysds.runtime.controlprogram.federated.FederatedData;
+import org.apache.sysds.runtime.controlprogram.federated.FederatedRange;
+import org.apache.sysds.runtime.controlprogram.federated.FederatedRequest;
+import org.apache.sysds.runtime.controlprogram.federated.FederatedResponse;
+import org.apache.sysds.runtime.instructions.InstructionUtils;
+import org.apache.sysds.runtime.instructions.cp.CPOperand;
+import org.apache.sysds.runtime.matrix.data.FrameBlock;
+import org.apache.sysds.runtime.matrix.operators.Operator;
+import org.apache.sysds.runtime.transform.encode.EncoderRecode;
+
+public class MultiReturnParameterizedBuiltinFEDInstruction extends ComputationFEDInstruction {
+	protected final ArrayList<CPOperand> _outputs;
+
+	private MultiReturnParameterizedBuiltinFEDInstruction(Operator op, CPOperand input1, CPOperand input2,
+		ArrayList<CPOperand> outputs, String opcode, String istr) {
+		super(FEDType.MultiReturnParameterizedBuiltin, op, input1, input2, outputs.get(0), opcode, istr);
+		_outputs = outputs;
+	}
+
+	public CPOperand getOutput(int i) {
+		return _outputs.get(i);
+	}
+
+	public static MultiReturnParameterizedBuiltinFEDInstruction parseInstruction(String str) {
+		String[] parts = InstructionUtils.getInstructionPartsWithValueType(str);
+		ArrayList<CPOperand> outputs = new ArrayList<>();
+		String opcode = parts[0];
+
+		if(opcode.equalsIgnoreCase("transformencode")) {
+			// one input and two outputs
+			CPOperand in1 = new CPOperand(parts[1]);
+			CPOperand in2 = new CPOperand(parts[2]);
+			outputs.add(new CPOperand(parts[3], Types.ValueType.FP64, Types.DataType.MATRIX));
+			outputs.add(new CPOperand(parts[4], Types.ValueType.STRING, Types.DataType.FRAME));
+			return new MultiReturnParameterizedBuiltinFEDInstruction(null, in1, in2, outputs, opcode, str);
+		}
+		else {
+			throw new DMLRuntimeException("Invalid opcode in MultiReturnBuiltin instruction: " + opcode);
+		}
+
+	}
+
+	@Override
+	public void processInstruction(ExecutionContext ec) {
+		// obtain and pin input frame
+		FrameObject fin = ec.getFrameObject(input1.getName());
+		String spec = ec.getScalarInput(input2).getStringValue();
+
+		Map<FederatedRange, FederatedData> fedMapping = fin.getFedMapping();
+
+		// first we use the spec to construct a meta frame which will provide us with info about the encodings
+		List<Pair<FederatedRange, Future<FederatedResponse>>> metaFutures = new ArrayList<>();
+		for(Map.Entry<FederatedRange, FederatedData> entry : fedMapping.entrySet()) {
+			Future<FederatedResponse> response = entry.getValue().executeFederatedOperation(new FederatedRequest(
+				FederatedRequest.FedMethod.ENCODE_META, spec, entry.getKey().getBeginDimsInt()[1] + 1), true);
+			metaFutures.add(new ImmutablePair<>(entry.getKey(), response));
+		}
+
+		// TODO support encodings other than recode
+		// the combined mappings for the frame columns (because we only support recode)
```

Comment:
I think it is important that we already think about the extensions. 
Since you implemented it for recode, how would you extend for others?
If the answer is complicated we should consider changing the code to make it less so.

Reply:
Yes, I refer to my comment above.